Fix #11698. Only call cholmod_start when CHOLMOD module is initialized

andreasnoack opened the pull request:
This avoids the memory leak. In consequence, introduce global `cholmod_common` array and restrict all
methods to `SuiteSparse_long` and remove the integer type parameter in CHOLMOD and SPQR types.

I don't think it is allowed to mix 32 and 64 bit integers in the same CHOLMOD session and since we only can have one CHOLMOD session we'll have to decide on a size on startup. Right now I've chosen `SuiteSparse_long`.

There is a second commit that uses gc tracked allocation functions in SuiteSparse. This makes the gc aware of the memory consumption of the objects allocated by the libraries in SuiteSparse. I couldn't find a `calloc` version of the gc tracked allocators and I don't know if this will be a problem. It appears to work fine.

andreasnoack committed:
Only call cholmod_start when CHOLMOD module is initiazed to avoid memory
leaks. In consequence, introduce global common array and restrict all
method to SuiteSparse_long.

Adjust CHOLMOD tests

andreasnoack commented:
@tkelman Which version of SuiteSparse is used in the AppVeyor builds?

@carnaval Would it make sense to add a `jl_gc_counted_calloc`?

tkelman commented:
Whatever's in the nightlies. 4.4.2 right now.

yuyichao commented:
@andreasnoack You mentioned that the library may call `realloc` on these pointers, Would the GC still be able to track the size of the allocation correctly then?

tkelman commented on base/sparse/cholmod.jl:
```diff
@@ -122,6 +133,15 @@ function __init__()
+    global const cmn = fill(0xff, common_size)
+    start(cmn)
+
+    cnfg = cglobal((:SuiteSparse_config, :libcholmod), Ptr{Void})
```
This symbol is in `libsuitesparseconfig`, at least with recent versions of suitesparse. I don't know where it is, if anywhere, on older versions. This may mean we need to bump the minimum suitesparse version requirement, cc @nalimilan.

KristofferC commented:
One of the test fails is a segfault in the garbage collector. Is it relevant to this PR?

tkelman commented:
I think that was the file-test problem, it was only on the "push" build of this branch, not the "pr" build of the merge commit. Rebasing against latest master might make it go away.

yuyichao commented:
Probably not. I believe that's because @andreasnoack haven't rebased his branch on master after https://github.com/JuliaLang/julia/pull/11704 yet and the `/push` test doesn't merge automatically

andreasnoack commented:
@tkelman Thanks for pointing to the right library. The linking seems to work differently on Windows and Linux so on Linux the symbol was found even when I pointed to the wrong library. I've fixed that now so hopefully it will work on Windows. We might still need to change the version requirement to be sure that the config struct exists. I'll look into that now.

@yuyichao I'm not completely sure. It is my first time to use the gc tracked allocators, but I'm using `jl_gc_counted_realloc_with_old_size` as `realloc` function pointer that seemed to be the approach in gmp.jl.

@KristofferC I think the segfault is happening somewhere else, but it's hard to tell. I've rebased so if any fixes have been made on master they should now be here as we'll.

timholy commented:
> You mentioned that the library may call realloc on these pointers, Would the GC still be able to track the size of the allocation correctly then?

Presumably it's good enough to have a rough estimate of size? The purpose is to decide when it's necessary to run GC again, and presumably the only penalty for getting it wrong is a performance hit. But the hit should be in proportion to the size of the error, so a ballpark estimate seems likely to be OK.

andreasnoack commented:
The test error on `/push` didn't go away after I restarted the tests. I've tried to reproduce it locally on 32 bit without luck. I'm wondering if there is anything special about the Travis machine.

KristofferC commented:
Hmm, aren't the test failure now a normal test fail in cholmod?

```
exception on 4: ERROR: LoadError: LoadError: test failed: [3.348981779926797 0.0 0.0 0.0 0.0
 0.357300512315947 3.298828959530392 0.0 0.0 0.0
 0.0 0.0 2.27165482875886 0.0 0.0
 0.0 0.0 0.0 2.020567640763385 0.0
 0.0 0.0 0.0 0.0 3.1125124222588907] == [1.8300223441058847 0.0 0.0 0.0 0.0
 0.6538679210986629 1.81626786557776 0.0 0.0 0.0
 0.0 0.0 1.507200991493457 0.0 0.0
 0.0 0.0 0.0 1.4214667216517538 0.0
 0.0 0.0 0.0 0.0 1.7642313970278645]
 in expression: CHOLMOD.Sparse(cholfact(Symmetric(A1pd,:L),shift=2)) == CHOLMOD.Sparse(cholfact(A1pd,shift=2))
```

andreasnoack commented:
It looks like it is, but I cannot reproduce it locally and the other 32 bit travis tests pass. Are you able to reproduce it locally?

tkelman commented:
I can reproduce the failure locally, from 64 bit Linux with the following `Make.user` to build a 32 bit Julia
```
override ARCH = i686
override MARCH = i686
override JULIA_CPU_TARGET = pentium4
```
You'll likely need to have multilib compilers installed, and comment out the libgit2 tests (it doesn't check very carefully that the openssl it's linking against is the same bitness as the compiler it's using), depending on your distribution. I have a Dockerfile that uses the same set of system libs as on the 32-bit Travis build that I can find a link to, or a proper 32 bit VM would also work.

KristofferC commented:
Is the failure due to some earlier corruption or does it fail if you only run the currently failing test?

andreasnoack commented:
@tkelman I've now tried with
 - 64 bit machine with 32 bit Ubuntu
 - 32 bit virtual machine with 32 bit Ubuntu
 - 64 bit virtual machine with 32 64 bit Ubuntu, but compiled with 32 bit libraries as you described

and still no test failures. Could you provide a link to your Dockerfile?

tkelman commented:
https://gist.githubusercontent.com/tkelman/63c2ce16dd2863cae17a/raw/ae7cc7ef82789d19eeb127db70fa2ff198aa694f/Dockerfile

How many cores are you running the tests on? Try `JULIA_CPU_CORES=1 ./julia test/runtests.jl core strings dict hashing tuple reduce reducedim intfuncs simdloop sparse` a few times. (I'm still cutting that list down.)

andreasnoack commented:
Thanks, but I'm not there yet. For some reason, I'm getting the error when building the Docker image
```jl
error during bootstrap:
LoadError(at "sysimg.jl" line 99: LoadError(at "regex.jl" line 5: LoadError(at "pcre.jl" line 28: UndefVarError(var=:ANCHORED))))
```

yuyichao commented:
That looks like the error I get from a old PCRE version or not cleaning up the base directory after PCRE upgrade.

tkelman commented:
Yeah, my mistake, remove `PCRE` from the for loop

andreasnoack commented:
Is it problem that I use Ubuntu 14.04? When I remove `PCRE` I get
```jl
/home/user/julia/deps/jlchecksum pcre2-10.10.tar.bz2
`which gtar 2>/dev/null || which tar 2>/dev/null` jxf pcre2-10.10.tar.bz2
tar (child): bzip2: Cannot exec: No such file or directory
tar (child): Error is not recoverable: exiting now
/bin/tar: Child returned status 2
/bin/tar: Error is not recoverable: exiting now
make[2]: *** [pcre2-10.10/configure] Error 2
make[2]: *** Waiting for unfinished jobs....
touch -c libuv/configure
warning: Remote branch 1.3-dev1 not found in upstream origin, using HEAD instead
touch -c utf8proc/Makefile
warning: Remote branch v0.22.2 not found in upstream origin, using HEAD instead
touch -c libgit2/CMakeLists.txt
make[2]: Leaving directory `/home/user/julia/deps'
make[1]: *** [julia-deps] Error 2
make[1]: Leaving directory `/home/user/julia'
make: *** [install] Error 2
```

tkelman commented:
Must be a really minimal base image, needs `bzip2` (maybe `bzip2:i386` ?) to be manually installed?

andreasnoack commented:
Thanks. That fixed it so now I've been able to build and test. However, still no CHOLMOD test errors here so I'll go to bed.

andreasnoack commented:
@tkelman I've now access to a Travis machine and I'm able to reproduce the test error. Have you been able to make the test fail with `JULIA_CPU_CORES=1`? So far I've only been able to do so with `JULIA_CPU_CORES=4`.

tkelman commented:
It fails more reliably for me when I set `JULIA_CPU_CORES=1`.

andreasnoack commented:
@tkelman Would you try to see if you can reproduce the test failure locally with my latest changes to the branch?

kshyatt commented:
@andreasnoack Is the failure only on Windows/Ubuntu? I've tried running your latest commits 4 times with `JULIA_CPU_CORES=1` and all tests have passed each time.

tkelman commented on test/sparsedir/cholmod.jl:
```diff
@@ -185,10 +190,6 @@ ACSC = sprandn(10, 10, 0.3) + I
 @test ishermitian(Sparse(Hermitian(complex(ACSC), :U)))
 
-# test Sparse constructor for c_Sparse{Tv,Ti} input( and Sparse*Sparse)
-B = CHOLMOD.Sparse(SparseMatrixCSC{Float64,Int32}(sprandn(48, 48, 0.1))) # A has Int32 indices
```
Would it be better to keep this test but change `Int32` to `SuiteSparse_long` ? Do we have `sparse(A) * sparse(B)` covered anywhere else?

andreasnoack replied:
This is covered in `sparsedir/sparse.jl`

tkelman replied:
That's not using the `CHOLMOD.Sparse` type though, is it?

andreasnoack replied:
`CHOLMOD.Sparse` multiplication is tested here 
https://github.com/JuliaLang/julia/blob/e432909e619174ad1ccb2dc778bb8d4fea12d7f8/test/sparsedir/cholmod.jl#L141-L169

tkelman commented on test/sparsedir/cholmod.jl:
```diff
@@ -256,19 +257,11 @@ unsafe_store!(puint, CHOLMOD.INTLONG, 3*div(sizeof(Csize_t), 4) + 5*div(sizeof(P
-    p = ccall((:cholmod_allocate_sparse, :libcholmod), Ptr{CHOLMOD.C_SparseVoid},
-        (Csize_t, Csize_t, Csize_t, Cint, Cint, Cint, Cint, Ptr{Void}),
-        1, 1, 1, true, true, 0, CHOLMOD.REAL, CHOLMOD.common(Int32))
-    @test isa(CHOLMOD.Sparse(p), CHOLMOD.Sparse{Float64,Int32})
```
Is this test worth keeping but with `Int32` replaced by `SuiteSparse_long` ?

andreasnoack replied:
This is covered by the other tests now the we don't support all integer sizes anymore.

tkelman commented:
The problem was only on 32 bit linux IIRC. I am having trouble reproducing it on the latest version of this branch, which is a good sign. 3 comments inline, on `$(INSTALL_NAME_CMD)` and deleted tests.

andreasnoack commented:
After a rebase I cannot reproduce the test failure anymore. It's a bit worrying that I don't know why it failed and why it is not failing anymore, but unless we can provoke the test to fail again soon I'm going to merge this to fix the memory leak in #11698

KristofferC commented:
Thanks for tackling this Andreas!

tkelman commented:
:cry: https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.6034/job/79sji16yujk9737x

andreasnoack commented:
The error is still there so I've reverted this.
